src/main/java/org/infinispan/persistence/mongodb/MongoDbStore.java

```diff
@@ -0,0 +1,193 @@
+    public static final String EXPIRY_TIME = "expiryTime";
+    public static final String CREATED = "created";
+    public static final String LAST_USED = "lastUsed";
+
+    private InitializationContext context;
+    private MongoClient mongoClient;
+    private MongoCollection<Document> collection;
+    private CacheToStoreConverter<K, V> cacheToStoreConverter;
+
+
+    @Override
+    public Set<Characteristic> characteristics() {
+        return EnumSet.of(Characteristic.SHAREABLE, Characteristic.BULK_READ, Characteristic.EXPIRATION);
+    }
+
+
+    @Override
+    public CompletionStage<Void> start(InitializationContext ctx) {
+        this.context = ctx;
+
+        MongoDbStoreConfiguration configuration = ctx.getConfiguration();
+        ConnectionConfiguration connectionConfiguration = configuration.connection();
+
+        this.cacheToStoreConverter = configuration.converter() == JSON
+                ? new JsonCacheToStoreConverter<>(ctx, ctx.getPersistenceMarshaller())
+                : new BinaryCacheToStoreConverter<>(ctx, ctx.getPersistenceMarshaller());
+
+        String connectionUri = connectionConfiguration.uri();
+        ConnectionString connString = new ConnectionString(connectionUri);
+        MongoClientSettings settings = MongoClientSettings.builder()
+                .applyConnectionString(connString)
+                .retryWrites(true)
+                .build();
+        this.mongoClient = MongoClients.create(settings);
+
+        String databaseName;
+        if (connString.getDatabase() != null) {
+            databaseName = connString.getDatabase();
+        } else {
+            databaseName = connectionConfiguration.database();
+        }
+        String collectionName = connectionConfiguration.collection();
+        MongoDatabase database = mongoClient.getDatabase(databaseName);
+        this.collection = database.getCollection(collectionName);
+
+        return CompletableFuture.completedFuture(null);
+    }
+
+    @Override
+    public CompletionStage<Void> stop() {
+        mongoClient.close();
+        return CompletableFuture.completedFuture(null);
+    }
+
+
+    @Override
+    public CompletionStage<Long> size(IntSet segments) {
+        return Single.fromPublisher(collection.countDocuments()).toCompletionStage();
+    }
+
+    @Override
+    public CompletionStage<Long> approximateSize(IntSet segments) {
+        return size(segments);
+    }
+
+
+    @Override
+    public CompletionStage<Boolean> containsKey(int segment, Object key) {
+        return load(segment, key).thenApply(Objects::nonNull);
+    }
+
+    @Override
+    public CompletionStage<MarshallableEntry<K, V>> load(int segment, Object key) {
+        long now = context.getTimeService().wallClockTime();
+        Object storeKey = cacheToStoreConverter.toStoreKey(key);
+        Bson filter = and(eq("_id", storeKey), or(gt(EXPIRY_TIME, now), lte(EXPIRY_TIME, -1L)));
+
+        return Maybe.fromPublisher(collection.find(filter))
```

Comment:
I personally would prefer something like `Flowable.fromPublisher(collection.find(filter)).singleStage()` to ensure we only ever receive a single entry from the Publisher, but I am fine with the way it is. TBH, I am surprised that `Maybe.fromPublisher` allows for more than 1 entry returned from a Publisher but `Single.fromPublisher` throws an exception.

And I know that since we are using _id that it shouldn't return more than one, but it is more protection of putting in a bug we don't expect.

Reply:
From the doc, `Maybe.fromPublisher` should emit "the very first item", so (unless I'm misunderstanding something) it shouldn't be a problem.

Correct?

Reply:
That is fine, I was more talking about the case when the publisher publishes more than one item, but this is a minor thing that is fine if not changed.